L1Trigger/TrackerTFP/src/SeedFilter.cc

```diff
@@ -0,0 +1,242 @@
+          const double cotGlobal = cot + setup_->sectorCot(eta);
+          const double zT = zT_.digi(z - cot * r);
+          const double z0 = zT - cot * setup_->chosenRofZ();
+          if (abs(z0) > setup_->beamWindowZ() + cot_.base() * setup_->chosenRofZ())
+            continue;
+          if (abs(zT) > dZT / 2. + zT_.base() / 2.)
+            continue;
+          //if (!cot_.inRange(cot))
+            //continue;
+          const vector<int>& layerEncoding = layerEncoding_->layerEncoding(eta, zT_.toUnsigned(zT_.integer(zT)), cot_.toUnsigned(cot_.integer(cot)));
+          vector<StubSF*> stubsSF;
+          stubsSF.reserve(track.size());
+          for (StubMHT* stub : track) {
+            const double sr = stub->r() + dT;
+            const double sz = stub->z();
+            const double chi = sz - (zT + sr * cot);
+            const double dZ = zT_.base() + cot_.base() * abs(sr) + setup_->dZ(stub->ttStubRef(), cotGlobal);
+            if (abs(chi) < dZ / 2.) {
+              int layer = distance(layerEncoding.begin(), find(layerEncoding.begin(), layerEncoding.end(), layerId(stub)));
+              if (layer >= setup_->numLayers())
+                layer = setup_->numLayers() - 1;
+              const double v = setup_->v1(stub->ttStubRef(), cotGlobal);
+              const double cov = pow(cot_.base() * stub->r(), 2) + pow(zT_.base(), 2);
+              const double chi2 = pow(chi, 2) / (v + cov);
+              stubsSF_.emplace_back(*stub, layer, zT_.integer(zT), cot_.integer(cot), chi2);
+              stubsSF.push_back(&stubsSF_.back());
+            }
+          }
+          set<int> ids;
+          for (StubSF* stub : stubsSF)
+            ids.insert(stub->layer());
+          if ((int)ids.size() >= setup_->sfMinLayers())
+            seedTracks.push_back(stubsSF);
+        }
+        if (seedTracks.empty())
+          continue;
+        seedTracks.erase(remove_if(seedTracks.begin(), seedTracks.end(), kill), seedTracks.end());
+        stable_sort(seedTracks.begin(), seedTracks.end(), lessSkippedLayers);
+        stable_sort(seedTracks.begin(), seedTracks.end(), smallerChi2);
+        stable_sort(seedTracks.begin(), seedTracks.end(), moreLayers);
+        const vector<StubSF*>& stubs = seedTracks.front();
+        copy(stubs.begin(), stubs.end(), back_inserter(output));
+      }
+      if (enableTruncation_ && ((int)output.size() > setup_->numFrames())) {
+        const auto limit = next(output.begin(), setup_->numFrames());
+        lostStream.reserve(distance(limit, output.end()));
+        transform(limit, output.end(), back_inserter(lostStream), [](StubSF* stub){ return stub->frame(); });
+        output.erase(limit, output.end());
+      }
+      acceptedStream.reserve(output.size());
+      transform(output.begin(), output.end(), back_inserter(acceptedStream), [](StubSF* stub){ return stub->frame(); });
+    }
+  }
+
+  //
+  int SeedFilter::layerId(StubMHT* stub) const {
+    int layer = stub->layer();
```

Comment:
Add comment "Convert from encoded layer ID to layer ID (barrel layer=1-6, endcap disk=11-15)